lightgbm/src/main/scala/com/microsoft/ml/spark/lightgbm/dataset/DatasetAggregator.scala

```diff
@@ -0,0 +1,557 @@
+import com.microsoft.ml.spark.lightgbm.swig.{BaseSwigArray, ChunkedArray, DoubleChunkedArray, DoubleSwigArray,
+  FloatChunkedArray, FloatSwigArray, IntChunkedArray, IntSwigArray}
+import com.microsoft.ml.spark.lightgbm.dataset.DatasetUtils.{addFeaturesToChunkedArray, addGroupColumnRow,
+  addInitScoreColumnRow, getArrayType, getRowAsDoubleArray}
+import com.microsoft.ml.spark.stages.LocalAggregator
+import org.apache.spark.ml.linalg.{DenseVector, SparseVector}
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.types.StructType
+
+import scala.collection.mutable.ListBuffer
+
+private[lightgbm] object ChunkedArrayUtils {
+  def copyChunkedArray[T: Numeric](chunkedArray: ChunkedArray[T],
+                                   mainArray: BaseSwigArray[T],
+                                   threadRowStartIndex: Long,
+                                   chunkSize: Long): Unit = {
+    val num = implicitly[Numeric[T]]
+    val defaultVal = num.fromInt(-1)
+    // Copy in parallel on each thread
+    // First copy full chunks
+    val chunkCount = chunkedArray.getChunksCount() - 1
+    for (chunk <- 0L until chunkCount) {
+      for (inChunkIdx <- 0L until chunkSize) {
+        mainArray.setItem(threadRowStartIndex + chunk * chunkSize + inChunkIdx,
+          chunkedArray.getItem(chunk, inChunkIdx, defaultVal))
+      }
+    }
+    // Next copy filled values from last chunk only
+    val lastChunkCount = chunkedArray.getLastChunkAddCount()
+    for (lastChunkIdx <- 0L until lastChunkCount) {
+      mainArray.setItem(threadRowStartIndex + chunkCount * chunkSize + lastChunkIdx,
+        chunkedArray.getItem(chunkCount, lastChunkIdx, defaultVal))
+    }
+  }
+}
+
+private[lightgbm] trait AggregatedColumns
+
+private[lightgbm] abstract class BaseChunkedColumns(columnParams: ColumnParams,
+                                                    chunkSize: Int) extends AggregatedColumns {
+  val labelsChunkedArray = new FloatChunkedArray(chunkSize)
+  val weightChunkedArrayOpt = columnParams.weightColumn.map {
+    _ => new FloatChunkedArray(chunkSize)
+  }
+  val initScoreChunkedArrayOpt = columnParams.initScoreColumn.map {
+    _ => new DoubleChunkedArray(chunkSize)
+  }
+  val groupColumnValues: ListBuffer[Row] = new ListBuffer[Row]()
+
+  var rowCount = 0
+
+  def addRow(): Unit = {
+    rowCount += 1
+  }
+
+  def release(): Unit = {
+    // Clear memory
+    labelsChunkedArray.delete()
+    weightChunkedArrayOpt.foreach(_.delete())
+    initScoreChunkedArrayOpt.foreach(_.delete())
+  }
+}
+
+private[lightgbm] final class SparseChunkedColumns(columnParams: ColumnParams,
+                                                   chunkSize: Int) extends BaseChunkedColumns(columnParams, chunkSize) {
+  var indexesChunkedArray = new IntChunkedArray(chunkSize)
+  var valuesChunkedArray = new DoubleChunkedArray(chunkSize)
+  var indptrChunkedArray = new IntChunkedArray(chunkSize)
+
+  var numCols = 0
```

Comment:
likewise here

Reply:
I made it protected, it is used in DatasetAggregator class so I added another method getNumCols()